src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -46,8 +49,8 @@ import           System.FS.BlockIO.API (HasBlockIO)
 --
 -- Since we always resolve all entries of the same key in one go, there is no
 -- need to store incompletely-resolved entries.
-data Merge m h = Merge {
-      mergeType       :: !MergeType
+data Merge t m h = Merge {
+      mergeType       :: !t
```

Comment:
I wonder if it would be a good idea here to store t and the result of `toMergeType mt` upon construction.

As it is, in `steps` we have to deal with all three cases anyway, but we incur a call via a class dict, and that adds a class dict to many of the merge functions below (which propagate upwards).

We could just have the class constraint on new merge, and then store the result and avoid class dicts being passed later on.

Or alternatively, don't store `t` at all if we don't need to retrieve it, just store the `toMergeType mt :: MergeType`. But keep the type param on `Merge`.

Reply:
Same. Now redundant.